File: VMLib/InstructionRegister.cs
namespace VMLib {
    internal class InstructionRegister : Component {
        public ushort OpCode { get; private set; }
        public ushort Arg1 { get; private set; }
        public ushort Arg2 { get; private set; }
        public ushort Arg3 { get; private set; }

        public InstructionRegister(VMCore core, Bus instBus, Bus dataBus) : base(core, nameof(InstructionRegister)) {
            Commands.Add(new MicrocodeCommand("Instruction Register In", () => {
                OpCode = (ushort)((instBus.Read() >> 9) & 0b1111111); // 7*1, 9*0
                Arg1 = (ushort)((instBus.Read() >> 6) & 0b111); // 7*0, 3*1, 6*0
                Arg2 = (ushort)((instBus.Read() >> 3) & 0b111); // 7*0, 3*0, 3*1, 3*0
                Arg3 = (ushort)((instBus.Read() >> 0) & 0b111); // 7*0, 6*0, 3*1
                OnPropertyChanged("OpCode");
                OnPropertyChanged("Arg1");
                OnPropertyChanged("Arg2");
                OnPropertyChanged("Arg3");
            }) { Changes = { this }, Depends = { instBus } });

            Commands.Add(new MicrocodeCommand("Intermediate", () => {
                dataBus.Write(Arg3);
            }) { Changes = { dataBus }, Depends = { this } });

            Commands.Add(new MicrocodeCommand("Long Intermediate", () => {
                dataBus.Write((ushort)(Arg2 << 3 | Arg3));
            }) { Changes = { dataBus }, Depends = { this } });
        }
    }
}

File: microasm/src/emulator/runtime/instructionRegister.c
#include "emulator/runtime/instructionRegister.h"
#include "shared/memory.h"
#include "shared/platform.h"
#include <stdlib.h>

// register to store the opcode and arguments of the currently
// executing instruction

// context for the microcode command
typedef struct instRegCtx {
    // which bus to read from
    unsigned int bus;
    // which register to set the values in
    instructionRegister* reg;
} instRegCtx;

// set the values in the intruction register
static void instRegSet(VMCore* core, void* vctx) {
    // cast context to correct type
    instRegCtx* ctx = vctx;

    // should not read from invalid busses
    // propper error handling todo
    if(!core->buss[ctx->bus].isValid) {
        cErrPrintf(TextRed, "Invalid bus read - ireg");
        exit(1);
    }

    uint16_t inst =  core->buss[ctx->bus].value;
    ctx->reg->value = inst;
    ctx->reg->opcode = (inst >> 9) & 0x7F; // first 7 bits
    ctx->reg->arg1 =   (inst >> 6) & 0x7;  // next 3 bits
    ctx->reg->arg2 =   (inst >> 3) & 0x7;  // next 3 bits
    ctx->reg->arg3 =   (inst >> 0) & 0x7;  // next 3 bits
    ctx->reg->arg12 = (ctx->reg->arg1 << 0x7) + ctx->reg->arg2;  // bits 8-13
    ctx->reg->arg123 = (ctx->reg->arg1 << 0x7) + 
        (ctx->reg->arg2 << 0x7) + ctx->reg->arg3; // bits 8-16
}

unsigned int instRegInit(VMCore* core, instructionRegister* reg, unsigned int instBus) {
    // context needs to live for life of vm
    instRegCtx* ctx = ArenaAlloc(sizeof(instRegCtx));
    ctx->bus = instBus;
    ctx->reg = reg;

    // set up command
    ADD_COMMAND(instRegSet, ctx, &core->buss[instBus]);
    CHANGES(reg);

    // return the command's id 
    return core->commandCount - 1;
}